src/plugins/ui_actions/public/context_menu/build_eui_context_menu_panels.tsx

```diff
@@ -231,30 +235,85 @@ export async function buildContextMenuForActions({
   // Any additional items are hidden behind a "more" item
   wrapMainPanelItemsIntoSubmenu(panels, 'mainMenu');
 
+  // This will be used to store items that eventually are placed into the
+  // mainMenu panel. Specifying a category allows for placing groups into the
+  // mainMenu so they appear without the separator between them.
+  const categories = {};
+
   for (const panel of Object.values(panels)) {
     // If the panel is a root-level panel, such as the parent of a group,
     // then create mainMenu item for this panel
     if (panel._level === 0) {
-      // Add separator with unique key if needed
-      if (panels.mainMenu.items.length) {
-        panels.mainMenu.items.push({ isSeparator: true, key: `${panel.id}separator` });
-      }
+      // If a category is specified, store either a link to the panel or the
+      // item within. We will deal with it after looping through all panels.
+      if (panel._category) {
```

Comment:
So many nested if conditions seems like a code smell, there also seems to be a decent amount of similar code, can this be refactored to something more manageable?

Reply:
I was able to do a `continue;` to reduce the `if` statements a bit, which helps, but I'm not sure there's a good way trim it down further. I hear you though, there are similar checks within the top-level if...I just find it makes the most sense to break apart the logic in the way it is currently structured.

Reply:
Thanks, yeah this is a bit more readable but a refactor here can be done in future. I'm good with this for now :)